R/ebird_conn.R

```diff
@@ -46,35 +49,38 @@ ebird_conn <- function(dataset = c("observations", "checklists"),
   
   # check for a cached connection
   conn <- mget("birddb", envir = birddb_cache, ifnotfound = NA)[["birddb"]]
-  if (inherits(conn, "DBIConnection")) {
-    if (DBI::dbIsValid(conn)) {
-      # if view not already in databse, create it
-      if (!dataset %in% DBI::dbListTables(conn)) {
-        DBI::dbSendQuery(conn, view_query)
-      }
-      return(conn)
-    } else {
-      # shut down invalid cached connection to allow re-connection
-      DBI::dbDisconnect(conn, shutdown = TRUE)
-    }
-  }
   
-  # TODO: consider persisting connection on disk to avoid recreating view
-  conn <- DBI::dbConnect(drv = duckdb::duckdb(), ebird_db_dir())
+  # Disconnect if it's an invalid connection (expired in cache)
+  if ( db_is_invalid(conn) )  
+    conn <- DBI::dbDisconnect(conn, shutdown = TRUE)
+  
+  # We don't have a valid cached connection, so we must create one! 
+  if (!inherits(conn, "DBIConnection")){
+    conn <- DBI::dbConnect(drv = duckdb::duckdb(), ebird_db_dir())  
+  }
   
-  # set memory limit
+  # enforce memory limit
   # TODO: temp approach for testing, improve approach in production
   DBI::dbExecute(conn = conn, 
                  paste0("PRAGMA memory_limit='", memory_limit, "GB'"))
+
+  # create the view if does not exist
+  if (!dataset %in% DBI::dbListTables(conn)){
+    DBI::dbSendQuery(conn, view_query)
+  }
   
-  # create view to parquet file
-  DBI::dbSendQuery(conn, view_query)
-  
-  # cache the connection
+  # (re)-cache the connection
   if (cache_connection) {
     assign("birddb", conn, envir = birddb_cache)
   }
-  return(conn)
+  
+  conn
+}
+
+## 
+db_is_invalid <- function(conn) {
```

Comment:
precisely what I wanted to write too!  The trouble is that if it `conn == NA`, calling `inherits(conn, "DBIConnection") && !DBI::dbIsValid(conn)` will error because it will try and evaluate both clauses before evaluating `&&`, and you cannot call `DBI::dbIsValid(conn)` on a `conn` that isn't even a connection object.  

Jenny taught me that this problem where you need to check a class (e.g. with `inherits`) before trying to run something on it is really better dealt with by using S3 methods, i.e. you would have methods  `db_is_invalid.DBIConnection()` and `db_is_invalid.logical()`.  That way, when you call `db_is_invalid()` the dispatch would call the appropriate method for the type, rather than needing any `if` statement.  We could go that way, though I don't recall if `db_is_invalid.DBIConnection()` works for S4 objects without some extra wrangling; my object-oriented-programming is pretty rusty.

Reply:
oh, weird, tbh I totally thought R used short circuit boolean operators for the non-vectorized versions (`&&/||`), e.g. this seems like it works for me:

```
conn <- NA
inherits(conn, "DBIConnection") && !DBI::dbIsValid(conn)
```

Although the vectorized version does always evaluate all components, e.g. this fails for me:

```
conn <- NA
inherits(conn, "DBIConnection") & !DBI::dbIsValid(conn)
```

But if that doesn't solve it, seems fine to me to just leave it as is, it's not terribly confusing.

Reply:
oh, that's great! didn't know this!